packages/compat/src/fixup-rules.js

```diff
@@ -0,0 +1,256 @@
+ * @type {WeakSet<FixupPluginDefinition>}
+ */
+const fixedUpPlugins = new WeakSet();
+
+//-----------------------------------------------------------------------------
+// Exports
+//-----------------------------------------------------------------------------
+
+/**
+ * Takes the given rule and creates a new rule with the `create()` method wrapped
+ * to provide the missing methods on the `context` object.
+ * @param {FixupRuleDefinition|FixupLegacyRuleDefinition} ruleDefinition The rule to fix up.
+ * @returns {FixupRuleDefinition} The fixed-up rule.
+ */
+export function fixupRule(ruleDefinition) {
+	// first check if we've already fixed up this rule
+	if (fixedUpRuleReplacements.has(ruleDefinition)) {
+		return fixedUpRuleReplacements.get(ruleDefinition);
+	}
+
+	const isLegacyRule = typeof ruleDefinition === "function";
+
+	// check to see if this rule definition has already been fixed up
+	if (!isLegacyRule && fixedUpRules.has(ruleDefinition)) {
+		return ruleDefinition;
+	}
+
+	const originalCreate = isLegacyRule
+		? ruleDefinition
+		: ruleDefinition.create.bind(ruleDefinition);
+
+	const ruleCreate = context => {
+		// if getScope is already there then no need to create old methods
+		if ("getScope" in context) {
+			return originalCreate(context);
+		}
+
+		const sourceCode = context.sourceCode;
+		let currentNode = sourceCode.ast;
+
+		const newContext = Object.assign(Object.create(context), {
+			parserServices: sourceCode.parserServices,
+
+			/*
+			 * The following methods rely on the current node in the traversal,
+			 * so we need to add them manually.
+			 */
+			getScope() {
+				return sourceCode.getScope(currentNode);
+			},
+
+			getAncestors() {
+				return sourceCode.getAncestors(currentNode);
+			},
+
+			getDeclaredVariables() {
+				return sourceCode.getDeclaredVariables(currentNode);
+			},
+
+			markVariableAsUsed(variable) {
+				sourceCode.markVariableAsUsed(variable, currentNode);
+			},
+		});
+
+		// add passthrough methods
+		for (const [
+			contextMethodName,
+			sourceCodeMethodName,
+		] of removedMethodNames) {
+			newContext[contextMethodName] =
+				sourceCode[sourceCodeMethodName].bind(sourceCode);
+		}
+
+		// freeze just like the original context
+		Object.freeze(newContext);
+
+		/*
+		 * Create the visitor object using the original create() method.
+		 * This is necessary to ensure that the visitor object is created
+		 * with the correct context.
+		 */
+		const visitor = originalCreate(newContext);
+
+		/*
+		 * Wrap each method in the visitor object to update the currentNode
+		 * before calling the original method. This is necessary because the
+		 * methods like `getScope()` need to know the current node.
+		 */
+		for (const [methodName, method] of Object.entries(visitor)) {
+			// node is the second argument for code path methods
+			if (methodName.startsWith("on")) {
+				visitor[methodName] = (...args) => {
+					currentNode = args[1];
```

Comment:
In case of `onCodePathSegmentLoop`, current node is `args[2]`.

Reply:
Got it.